### ISSUE TITLE
Add retry with exponential backoff

### ISSUE THREAD

**Issue by m90** [m90/go-deferred#1]
It would be nice if the library would know how to use [Exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) when retrying the handler creation.

Ideally a new option like `deferred.WithExponentialBackoff(time.Second)` would be created that is then passed in instead of `deferred.WithRetryAfter(time.Second)`.

**Comment by princerachit** [comment]
I want to work on this issue. Can you completely describe the method signature and behaviour of the handler?

What does the argument to `WithExponentialBackoff(arg time.Second)` mean? Does it mean that back-off starts with `arg` seconds and then `2*arg` and so on?

What would be hard limit on back-off? It should not keep on increasing by a factor of 2, we need to set some limit. Like `20000s` or `400000s`?

**Comment by m90** [comment]
I would have expected to be the argument the initial value for the first retry. The following retries would be determined by the algorithm (instead of simple constant repetition like now). As for the hard limit, there currently is `deferred.WithTimeoutAfter(timeout time.Duration)` which I would expect to still apply in this case.

That being said, this API is not set in stone and I would be happy if anyone would come up wiith a different solution that is user-friendly.

Maybe it's also a good idea to look at existing packages that implement the algorithm in Go and other languages.

**Comment by m90** [comment]
@princerachit before starting to work on this you might want to check with @ananichev (who has foreked this repo just recently) if they are currently already working on this.

**Comment by princerachit** [comment]
@m90 do we expect retryAfter to go away? Or there is a precedence that we want to set here?

Also @ananichev are you already working on this?

**Comment by m90** [comment]
I'd expect `WithRetryAfter` and `WithExponentialBackoff` to be mutually exclusive, i.e. if the user passes both, `NewHandler` should probably return an error upfront (although that'd result in a breaking API change, yet I think that's ok in this case). Defining a precedence would be confusing, at least to me.

**Comment by princerachit** [comment]
I am waiting for @ananichev 's reply

**Comment by m90** [comment]
@princerachit Cool, thanks for helping with finding out about that specs in any case :100:!

**Comment by ananichev** [comment]
@princerachit Yes, I started working on it, sorry that did not mention it in comments

**Comment by m90** [comment]
Closed via #2
